### PR TITLE
[WIP] General polymer implementation

### PR DESCRIPTION
**Pull request by connordavel**
More info and simple write-up/motivation can be found [here](https://docs.google.com/document/d/1LB1-9RpNWJ-R6wlNjJer5vlFB_LF2cNUhD9xS5QKq-c/edit). This PR adds custom substructure functionality to allow Users to modify the substructure dictionary and load custom polymers, functionalized proteins, or protein/polymer systems. Supersedes #1600. 

Checklist:
- [X] Input Handling for custom substructures and integration with existing substructure dictionary
- [X] Error handling for ambiguous chemical assignments from custom substructures. If custom substructures disagree with eachother or the existing library, raise exception
- [x] input validation for custom substructures that adhere to the following general string format: "...-[#<atomic num>D<explicit connect>+<formal charge>]-..."
- [x] simple testing on a few simple polymers and modified protein
- [x] testing to ensure input validation and exceptions

A short testing script with examples can be found [here.](https://github.com/connordavel/Minimum-Polymer-Tests)

### CODE REVIEW

File: openff/toolkit/utils/rdkit_wrapper.py
Comment (j-wags) [inline openff/toolkit/utils/rdkit_wrapper.py:992]: This is no longer true
```suggestion
                )  
```

File: utilities/make_substructure_dict/_make_chemical_substructures.py
Comment (j-wags) [inline utilities/make_substructure_dict/_make_chemical_substructures.py:119]: I'm not sure I see the purpose of this section. Is it redundant with [this](https://github.com/openforcefield/openff-toolkit/blob/f84ca0e1955f151cd9c842fbd9bb144f486ef29c/utilities/make_substructure_dict/_cif_to_substructure_dict.py#L610) or does it serve a different purpose?

**Review by j-wags — COMMENTED** [review]
Thanks for pushing this forward. I'll need to come back to figure out which whole-molecule loading tests to include but I've put a few comments/questions here. Could you also:
* Push the updated file to `openff/toolkit/data/proteins/aa_residues_substructures_with_caps.json` as well
* (optional) Add tests for the substructure validation code, which exercise all possible failure modes.

File: openff/toolkit/utils/rdkit_wrapper.py
Comment (connordavel) [reply openff/toolkit/utils/rdkit_wrapper.py:678]: Yes, I will remove this. Before I coded everything to retain the original substructure format, but if we are moving everything over, strict degree will always be used.

File: openff/toolkit/utils/rdkit_wrapper.py
Comment (connordavel) [reply openff/toolkit/utils/rdkit_wrapper.py:878]: This bit of code was added as a way to handle some tricky edge cases. For example, take carboxylate ions ,-C(-[O-])(=O). The code later one might error because the chemical information changes based on whether the graph match places the negative charge/double bond on one oxygen or another. 
In other cases of highly symmetric groups (including lots of aromatic cases) this code helps catch what would be chemical ambiguity errors. I'll add some comments and tests cases for this.

File: openff/toolkit/utils/rdkit_wrapper.py
Comment (connordavel) [reply openff/toolkit/utils/rdkit_wrapper.py:471]: I'll put a comment there. In the daylight smarts spec: "Note that atomic primitive H can have two meanings, implying a property or the element itself. [H] means hydrogen atom. [*H2] means any atom with exactly two hydrogens attached". Using atomic numbers instead just seemed cleaner in general, although not strictly necessary.

File: openff/toolkit/utils/rdkit_wrapper.py
Comment (connordavel) [reply openff/toolkit/utils/rdkit_wrapper.py:546]: The main issue with aromatic groups is that they can trick the code later into thinking that those substructures give chemically ambiguous assignments. In aromatic groups, the graph match will often return matches that maybe flip or rotate aromatic rings, so when a single bond is overwritten as a double, the expected behavior is to error. While the `_get_symmetrical_groups` function helps somewhat with aromatic groups, I think there may be edge cases that require the use of aromatic bonds (I think polyphenylene with all its aromatic rings is a good example of such a case). 

Could we make it work if we require (and somehow enforce) the use of the MDL aromaticity model on the substructures?

File: openff/toolkit/utils/rdkit_wrapper.py
Comment (connordavel) [reply openff/toolkit/utils/rdkit_wrapper.py:580]: This looks like an error. I'll fix this.

File: openff/toolkit/utils/rdkit_wrapper.py
Comment (connordavel) [reply openff/toolkit/utils/rdkit_wrapper.py:721]: We can. I originally coded this to not interfere with the original substructures/behavior of the function, but I'll start treating the amino acid substructures like custom substructures and run some tests!

File: utilities/make_substructure_dict/_make_chemical_substructures.py
Comment (connordavel) [reply utilities/make_substructure_dict/_make_chemical_substructures.py:119]: This section adds wildtype atoms where needed to satisfy the new spec, so I don't think it is redundant. However, I might see if this code can maybe run further up and will add a bit more documentation here to explain.

File: openff/toolkit/utils/rdkit_wrapper.py
Comment (j-wags) [reply openff/toolkit/utils/rdkit_wrapper.py:546]: I see your point and am thinking about it. Supporting aromatic bonds is unsatisfying/dangerous on a fundamental level - the information content of a graph molecule only needs two of (bond order, formal charge, aromaticity). If we have all three, the molecule is overdefined, and the sources of information become able to disagree with each other, leading to unnecessary edge cases that wouldn’t be there if only two of them were known.

What about requiring the “stable” (on disk) representations to be kekulized (so only integer bond order+formal charge), but then perceiving MDL aromaticity upon loading and during matching, and kekulizing the final loaded polymer at the end? Then we should get all the benefits of aromaticity in avoiding matching conflicts, but avoid using aromaticity in our input or outputs.

Comment (connordavel) [reply openff/toolkit/utils/rdkit_wrapper.py:546]: The "two degrees of freedom" argument makes a ton of sense actually, and I now agree aromaticity could be an issue. I will try out that solution that you mentioned. It should work since all of the aromaticity seen in my/Tim's examples use simple/MDL aromaticity. Thanks!

**Comment by j-wags** [comment]
(notes to self) To do:

* [ ] Get CI running
* [ ] (open new issue) Add a challenging test (like with resonance symmetry)
* [ ] (open new issue) Have `_custom_substructures` be `{str: List[str]}` instead of `{str: str}`?
* [ ] (open new issue) Rethink stereo check - may be nonsensical as written
* [ ] Update releasehistory

**Review by j-wags — COMMENTED** [review]
There’s something weird going on with CI but I’m generally optimistic about merging this by aug 1. There are one or two things I’ll open new issues about for followup but they don’t seem severe. Will get back to it tomorrow morning.

File: openff/toolkit/utils/rdkit_wrapper.py
Comment (connordavel) [reply openff/toolkit/utils/rdkit_wrapper.py]: I actually had some uncertainty in coding this originally. It was added because the original amino acid templates had stereochemistry. However, if we only get stereo from 3d, I agree that this is pointless. Would you support removing all mention of stereochem before the final assignment from 3d coordinates?

Comment (j-wags) [reply openff/toolkit/utils/rdkit_wrapper.py]: Gotcha - if you're also doubtful about this then I'll remove it. I've confirmed that we're still doing AssignStereochemistryFrom3D afterwards anyway, so this shouldn't cause a conflict.

**Comment by connordavel** [comment]
`(open new issue) Add a challenging test (like with resonance symmetry)`
Added 8fy3 which is a somewhat large protein that will test many amino acids and has a symmetrical ARG group that breaks error checking without symmetry detection.

**Comment by j-wags** [comment]
Thanks for giving that a shot, but I don't think 8fy3 adds anything that the [5tbm_complex_solv](https://github.com/openforcefield/openff-toolkit/blob/1689d993a386d2ca4b9e9f71ad82da5d007e63cd/openff/toolkit/_tests/test_topology.py#L696-L706) case didn't already test (it has multiple arginines and aromatic AAs). I need to think more about what a novel test case here could be  - Maybe natural rubber is a good case for this? Or even better would be a polymer with long conjugation (linear paths of alternating double bonds, possibly with some groups being charged)?

Anyway, let's handle those in another PR so we can get this merged. That PDB file is pretty big so it'd be best if we don't merge it with this PR - Could you revert that commit?

**Review by j-wags — APPROVED** [review]
Thanks for slogging through this one, @connordavel! This is really great, and I've got a good amount of confidence in it. I'll merge this once tests pass :-)

**Pull request merged**
